### PR TITLE
create TLS ServerParams once per server run; remove tlsServerParams from agent env (fixes functional agent client for chat)

### PR DESCRIPTION
**Pull request by efim-poberezkin**

### CODE REVIEW

File: src/Simplex/Messaging/Agent/Env/SQLite.hs [hunk 1]
```diff
@@ -86,8 +86,9 @@ newSMPAgentEnv cfg = do
   store <- liftIO $ createSQLiteStore (dbFile cfg) (dbPoolSize cfg) Migrations.app
   clientCounter <- newTVarIO 0
   randomServer <- newTVarIO =<< liftIO newStdGen
-  agentCredential <- liftIO $ loadServerCredential (agentPrivateKeyFile cfg) (agentCertificateFile cfg)
-  return Env {config = cfg, store, idsDrg, clientCounter, reservedMsgSize, randomServer, agentCredential}
+  serverCredential <- liftIO $ loadServerCredential (agentPrivateKeyFile cfg) (agentCertificateFile cfg)
+  let tlsServerParams = mkTLSServerParams serverCredential
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent/Env/SQLite.hs:90]: it makes sense to compose loadServerCredential and mkTLSServerParams into one function loadTLSServerParams that accepts key file and cert file paths (or getTLSServerParams) in Transport and call it from both Agent and Server

Comment (efim-poberezkin) [reply src/Simplex/Messaging/Agent/Env/SQLite.hs:90]: started doing it but liked it less..

File: src/Simplex/Messaging/Transport.hs [hunk 2]
```diff
@@ -197,12 +196,23 @@ startTCPClient host port = withSocketsDo $ resolve >>= tryOpen err
...
       connectTLS "client" getClientConnection clientParams sock
 
--- TODO non lazy
-loadServerCredential :: FilePath -> FilePath -> IO T.Credential
-loadServerCredential privateKeyFile certificateFile =
-  T.credentialLoadX509 certificateFile privateKeyFile >>= \case
-    Right cert -> pure cert
-    Left _ -> putStrLn "invalid credential" >> exitFailure
+mkTLSServerParams :: MonadUnliftIO m => FilePath -> FilePath -> m T.ServerParams
+mkTLSServerParams privateKeyFile certificateFile =
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:200]:
```suggestion
loadTLSServerParams :: MonadUnliftIO m => FilePath -> FilePath -> m T.ServerParams
loadTLSServerParams privateKeyFile certificateFile =
```

File: src/Simplex/Messaging/Transport.hs [hunk 3]
```diff
@@ -33,7 +33,7 @@ module Simplex.Messaging.Transport
     -- * Transport over TLS 1.3
     runTransportServer,
     runTransportClient,
-    loadServerCredential,
+    mkTLSServerParams,
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:36]:
```suggestion
    loadTLSServerParams,
```

File: src/Simplex/Messaging/Server/Env/STM.hs [hunk 4]
```diff
@@ -100,8 +100,8 @@ newEnv config = do
   s' <- restoreQueues queueStore `mapM` storeLog (config :: ServerConfig)
   let pk = serverPrivateKey config -- TODO remove
       serverKeyPair = (C.publicKey pk, pk)
-  serverCredential <- liftIO $ loadServerCredential (serverPrivateKeyFile config) (serverCertificateFile config)
-  return Env {config, server, queueStore, msgStore, idsDrg, serverKeyPair, storeLog = s', serverCredential}
+  tlsServerParams <- mkTLSServerParams (serverPrivateKeyFile config) (serverCertificateFile config)
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Server/Env/STM.hs:103]:
```suggestion
  tlsServerParams <- loadTLSServerParams (serverPrivateKeyFile config) (serverCertificateFile config)
```

File: src/Simplex/Messaging/Server/Env/STM.hs [hunk 5]
```diff
@@ -19,7 +19,7 @@ import Simplex.Messaging.Server.MsgStore.STM
 import Simplex.Messaging.Server.QueueStore (QueueRec (..))
 import Simplex.Messaging.Server.QueueStore.STM
 import Simplex.Messaging.Server.StoreLog
-import Simplex.Messaging.Transport (ATransport, SessionId, loadServerCredential)
+import Simplex.Messaging.Transport (ATransport, SessionId, mkTLSServerParams)
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Server/Env/STM.hs:22]:
```suggestion
import Simplex.Messaging.Transport (ATransport, SessionId, loadTLSServerParams)
```

File: src/Simplex/Messaging/Agent/Env/SQLite.hs [hunk 6]
```diff
@@ -86,8 +86,8 @@ newSMPAgentEnv cfg = do
   store <- liftIO $ createSQLiteStore (dbFile cfg) (dbPoolSize cfg) Migrations.app
   clientCounter <- newTVarIO 0
   randomServer <- newTVarIO =<< liftIO newStdGen
-  agentCredential <- liftIO $ loadServerCredential (agentPrivateKeyFile cfg) (agentCertificateFile cfg)
-  return Env {config = cfg, store, idsDrg, clientCounter, reservedMsgSize, randomServer, agentCredential}
+  tlsServerParams <- mkTLSServerParams (agentPrivateKeyFile cfg) (agentCertificateFile cfg)
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent/Env/SQLite.hs:89]:
```suggestion
  tlsServerParams <- loadTLSServerParams (agentPrivateKeyFile cfg) (agentCertificateFile cfg)
```

File: src/Simplex/Messaging/Agent/Env/SQLite.hs [hunk 7]
```diff
@@ -18,7 +18,7 @@ import Simplex.Messaging.Agent.Store.SQLite
 import qualified Simplex.Messaging.Agent.Store.SQLite.Migrations as Migrations
 import Simplex.Messaging.Client
 import qualified Simplex.Messaging.Crypto as C
-import Simplex.Messaging.Transport (loadServerCredential)
+import Simplex.Messaging.Transport (mkTLSServerParams)
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent/Env/SQLite.hs:21]:
```suggestion
import Simplex.Messaging.Transport (loadTLSServerParams)
```

File: src/Simplex/Messaging/Transport.hs [hunk 8]
```diff
@@ -197,12 +196,23 @@ startTCPClient host port = withSocketsDo $ resolve >>= tryOpen err
...
 
--- TODO non lazy
-loadServerCredential :: FilePath -> FilePath -> IO T.Credential
-loadServerCredential privateKeyFile certificateFile =
-  T.credentialLoadX509 certificateFile privateKeyFile >>= \case
-    Right cert -> pure cert
-    Left _ -> putStrLn "invalid credential" >> exitFailure
+mkTLSServerParams :: MonadUnliftIO m => FilePath -> FilePath -> m T.ServerParams
+mkTLSServerParams privateKeyFile certificateFile =
+  liftIO loadServerCredential <&> fromCredential
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:201]:
```suggestion
  liftIO $ fromCredential <$> loadServerCredential
```

File: src/Simplex/Messaging/Transport.hs [hunk 9]
```diff
@@ -68,7 +68,7 @@ import Data.ByteString.Char8 (ByteString)
 import qualified Data.ByteString.Char8 as B
 import qualified Data.ByteString.Lazy as BL
 import Data.Default (def)
-import Data.Functor (($>))
+import Data.Functor (($>), (<&>))
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:71]:
```suggestion
import Data.Functor (($>))
```

**Review by epoberezkin — APPROVED** [review]

**Pull request merged**